src/TesApi.Web/Management/Clients/HttpApiClient.cs

```diff
@@ -187,17 +196,56 @@ private async Task AddAuthorizationHeaderToRequestAsync(HttpRequestMessage reque
             }
         }
 
+        private async Task<string> GetOrRefreshAccessToken()
+        {
+            // here we are synchronizing access to the access token only, there is a still
+            // a small chance the call to GetTokenAsync won't be synchronized, and it will get call more than once, concurrently
+            // however the impact should be marginal - the cached token will be updated more than once. 
+            lock (lockObject)
```

Comment:
new SemaphoreSlim(1,1) works well here instead of lock since it has an async `WaitAsync()` method.  Can then add `.Release()` in a `Finally` block

https://learn.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim?view=net-7.0

Reply:
I considered but one concern was performance.  This method is a low-level call, in the http pipeline, and I wanted to avoid the performance impact of adding to the blocking time the I/O operation and the synchronization overhead of the semaphore, if possible.  In short, I wanted to synchronize r/w access to the property, which I can do with a simple lock.